flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -745,152 +863,184 @@ class ExprLowering {
+          auto prod = builder.create<mlir::MulIOp>(loc, scaled, sum);
+          total = builder.create<mlir::AddIOp>(loc, prod, total);
+          if (auto ext = std::get<0>(pair))
+            delta = builder.create<mlir::MulIOp>(loc, delta, ext);
+        } else if (auto *sval = subVal.getUnboxed()) {
+          auto val = builder.createConvert(loc, idxTy, *sval);
+          auto lb = builder.createConvert(loc, idxTy, getLB(arr, dim));
+          auto diff = builder.create<mlir::SubIOp>(loc, val, lb);
+          auto prod = builder.create<mlir::MulIOp>(loc, delta, diff);
+          total = builder.create<mlir::AddIOp>(loc, prod, total);
+          if (auto ext = std::get<0>(pair))
+            delta = builder.create<mlir::MulIOp>(loc, delta, ext);
+        } else {
+          TODO();
+        }
+        ++dim;
       }
       return builder.create<fir::CoordinateOp>(
           loc, refTy, base, llvm::ArrayRef<mlir::Value>{total});
     };
+    auto genArraySlice = [&](const auto &arr) -> mlir::Value {
+      // FIXME: create a loop nest and copy the array slice into a temp
+      // We need some context here, since we could also box as an argument
+      return builder.create<fir::UndefOp>(loc, refTy);
+    };
     return std::visit(
         Fortran::common::visitors{
-            [&](const Fortran::lower::SymIndex::Shaped &arr) {
-              return genShaped(arr, one);
-            },
-            [&](const Fortran::lower::SymIndex::FullDim &arr) {
+            [&](const Fortran::lower::SymbolBox::FullDim &arr) {
+              if (!inArrayContext() && isSlice(aref))
+                return genArraySlice(arr);
               return genFullDim(arr, one);
             },
-            [&](const Fortran::lower::SymIndex::CharShaped &arr) {
-              return genShaped(arr, arr.len);
-            },
-            [&](const Fortran::lower::SymIndex::CharFullDim &arr) {
+            [&](const Fortran::lower::SymbolBox::CharFullDim &arr) {
               return genFullDim(arr, arr.len);
             },
-            [&](const Fortran::lower::SymIndex::Derived &arr) {
+            [&](const Fortran::lower::SymbolBox::Derived &arr) {
               TODO();
               return mlir::Value{};
             },
             [&](const auto &) {
               TODO();
               return mlir::Value{};
             }},
-        si.v);
+        si.box);
   }
 
   // Return the coordinate of the array reference
-  mlir::Value gen(const Fortran::evaluate::ArrayRef &aref) {
+  Fortran::lower::ExValue gen(const Fortran::evaluate::ArrayRef &aref) {
     if (aref.base().IsSymbol()) {
       auto &symbol = aref.base().GetFirstSymbol();
       auto si = symMap.lookupSymbol(symbol);
       if (!si.hasConstantShape())
         return gen(si, aref);
-      mlir::Value base = gen(symbol);
-      auto &shape =
-          symbol.get<Fortran::semantics::ObjectEntityDetails>().shape();
+      auto box = gen(symbol);
```

Comment:
Why not using `Fortran::lower::ExValue gen(const Fortran::lower::SymbolBox &si, const Fortran::evaluate::ArrayRef &aref)` in case the shape is not constant ? Is it not handling such cases ?

Reply:
This is a trick. It actually calls `Fortran::lower::ExValue gen(const Fortran::lower::SymbolBox &si, const Fortran::evaluate::ArrayRef &aref)` when the shape is _NOT_ constant. So that case is handled.
The sequent is handling the _constant_ shape case.
Let me see if these cases can be merged together now.